### ISSUE TITLE
Support for Active Directory B2B Users?

### ISSUE THREAD

**Issue by imaksikkar** [OfficeDev/Office-Add-in-ASPNET-SSO#11]
I attempted to use this SSO feature when logged into Word as an Active Directory B2B (aka guest) user. This user was created in Active Directory using the following interface:

![image](https://user-images.githubusercontent.com/3792784/30389876-06dec0f6-9869-11e7-95bf-a4202af8f832.png)

The call to `Office.context.auth.getAccessTokenAsync` gave me the following error:

![image](https://user-images.githubusercontent.com/3792784/30389751-a28bcfc2-9868-11e7-90d5-261e8561dafc.png)

Does this mean that guest/B2B users are not supported for SSO with Office? And if so, is there planned support for this?

**Comment by TimMcMSFT** [comment]
Hi Dana, could you elaborate on the process? You created a new guest user in AAD. What type of external address did you use? Was this the same address that you used to sign into Office? The SSO feature is built upon Office's understanding of the signed in user. From a scenario perspective, I'd also like to understand how your solution or organization uses Guest Accounts with Office.

**Comment by imaksikkar** [comment]
Absolutely. I currently have two guest users in this Active Directory. One is an @gmail.com user that was forced to create a password with Microsoft upon accepting the invitation to join my test organization. The second is an @outlook.com (personal Microsoft account) user that is able to use their existing Microsoft credentials to access my test organization after accepting the invitation. 

I have tested with both of those users by logging them into Office and running the code that calls `Office.context.auth.getAccessTokenAsync` and both get the same error message that I pasted above. I did make sure that no other user was logged into Office except the guest user I was testing with.

My organization will be using these guest users to give external customers controlled access to specific sites in our SharePoint Online. So, for example, a guest user will be logged into Word and our add-in will hopefully be able to use this SSO-with-Office feature to acquire an access token for Graph so that we can query the specific site that the user has access to.

**Comment by TimMcMSFT** [comment]
Ok, thanks for the info Dana. My suspicion is that your build (I assume that it's [an Insiders Fast build](https://products.office.com/en-us/office-insider)) hasn't added support for Microsoft Accounts, yet. Our API Preview was only for Org Accounts, and MSAs are rolling out in preparation for the API's GA. If you have a build of Office after 8515, then the API should work without issue for MSAs.

**Comment by TimMcMSFT** [comment]
Update--it looks like 8530 or later is what you'll need.

**Comment by imaksikkar** [comment]
So I've updated Office to build 8530 and my error message for `Office.context.auth.getAccessTokenAsync` with a guest user has changed to this:

![image](https://user-images.githubusercontent.com/3792784/30444381-a1eb13c8-9937-11e7-90ef-16127101517c.png)

The error is identical for both the @gmail.com and @outlook.com users. I've also noticed that the place in the upper right hand corner of Word that shows the logged in user looks normal before the add-in SSO code runs but looks like this after it runs and gives that error:

![image](https://user-images.githubusercontent.com/3792784/30444486-04736efa-9938-11e7-925b-b9c177c30e87.png)

Not sure if that's related but it happens reliably right when the call to get the access token from Word fails. This also happens for both guest users.

**Comment by Rick-Kirkham** [comment]
One known cause of 13007 is that, contrary to our current documentation which we are working to update, you need to ask for "profile" permission in the Microsoft Graph Permissions section of the app registration. Then you need to redo the admin consent process. Please see if that helps.

**Comment by imaksikkar** [comment]
I have asked for and granted consent for "profile" in this app already. I was originally having an issue with 13007 for all users and adding the profile permission resolved that issue. This new issue only applies to guest users, my normal Active Directory admin user isn't having any problems.

**Comment by imaksikkar** [comment]
Wondering if there's any update on this? If this SSO-with-Office feature isn't supported for guest users then I'm planning to use the [MSAL library preview for JavaScript](https://github.com/AzureAD/microsoft-authentication-library-for-js) to get a Graph access token for the user. I'd prefer not to take this approach if it's possible to use the SSO solution for guest users though because I'd rather not force the user to enter their Microsoft credentials if they are already logged into Word.

**Comment by TimMcMSFT** [comment]
Dana, when you say that you granted consent--did you have the Guest accounts do this individually? Or are you speaking about an admin consent for the organization? I'm trying to understand if the issue here could be that the original 13007 error, which you resolved for the Tenant, still occurs for Guests because consent works differently for them.

**Comment by imaksikkar** [comment]
I mean that I granted admin consent for the organization. That's an interesting idea. I've never seen a consent prompt for a guest user when using this feature. Is that something I could trigger by changing the `forceConsent` parameter to true in the `Office.context.auth.getAccessTokenAsync()` call?

**Comment by imaksikkar** [comment]
As an update, I did try changing `forceConsent` to true but that change is causing me to get the following error for all users I try it with:

![image](https://user-images.githubusercontent.com/3792784/30764547-5370d2f4-9fa0-11e7-99d9-f4b8ea1132f6.png)

Is that not the correct way to ensure that my users get prompted to consent? I have yet to see the consent screen for a user in any of my testing or dev so far with this feature other than the admin consent

**Comment by TimMcMSFT** [comment]
Hi Dana, assuming that you're still in the development phase, i.e. your add-in hasn't made it to the Office Store, then our Consent experience cannot help (yet). In order to do user consent, you'll need to go directly against Azure's authentication endpoint with the guest users, something like:
```
https://login.microsoftonline.com/common/oauth2/v2.0/authorize?
client_id={your cliend it}&
response_type=id_token&
redirect_uri=https://login.live.com/oauth20_desktop.srf&
scope=openid%20profile%20{any scopes that you need}&
response_mode=fragment&state=12345&nonce=678910&
prompt=consent
```
Once your add-in is in the store, Office will handle the Consent step for these guest accounts.

**Comment by imaksikkar** [comment]
I filled in that url with the application ID for my app, the scopes that I need and with a valid redirect uri for my application and navigated to it in a private browsing window. I logged in as my guest user (@outlook.com) and consented to the scopes it showed me. Everything looked good and it redirected me correctly with an ID token to where my app is running locally. I restarted my application, made sure I was logged in as the guest user only and tried to call `Office.context.auth.getAccessTokenAsync()` again but my error didn't change. It's still giving me a 13007 error.

It seemed like the consent was successful but I'm still getting the error. I ran Fiddler during the call that's failing and didn't see any traffic to login.windows.net or anywhere else relevant. Is it possible that it's blowing up with that error before it even calls out to the token endpoint?

**Comment by imaksikkar** [comment]
Update because I've made it farther on this issue:

From discussions in a ticket with Microsoft, it sounds like the user consent screen is not shown unless the add-in is hosted in the Office store. Because of that, I needed to login as my guest user and manually navigate to the consent prompt here:

`https://login.microsoftonline.com/common/oauth2/v2.0/authorize?client_id=CLIENT_ID&response_type=code&redirect_uri=REDIRECT_URI&scope=ADDITIONAL_SCOPES%20offline_access%20openid%20profile&response_mode=query&state=12345`

After consenting as the guest user, I make it past the call to `Office.context.auth.getAccessTokenAsync()` and get an access token. However, I don't get the necessary `access_as_user` scope claim and therefore can't exchange my Office token for a Graph token. When inspecting the Office token, I see that the `scp` property is present but is set to empty string.

Is there a reason that the Office token for a guest user would be invalid to exchange for a Graph token?
